Introduce utility to generate network genesis data

ralexstokes opened the pull request:
### What was wrong?

We need a single point of entry to generate a (self-contained) piece of data that describes a network genesis for both the beacon node and the validator client.

### How was it fixed?

Introduce an `Application` to accomplish this task; will deprecate the `Interop` component in a subsequent PR. NOTE: The genesis data is written in a custom JSON format that only Trinity nodes will know how to handle.

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://api.time.com/wp-content/uploads/2020/03/coronavirus-dog.jpg?quality=85&w=1024&h=628&crop=1)

pipermerriam commented on eth2/beacon/tools/builder/initializer.py:
```diff
@@ -54,6 +106,64 @@ def create_keypair_and_mock_withdraw_credentials(
+    for i in range(validator_count):
+        private_key = generate_privkey_from_index(i)
+        public_key = bls.privtopub(private_key)
+        key_pairs[public_key] = private_key
```
We have `eth_utils.to_dict` which allows you to do:

```python
@to_dict
def create_key_pairs_for(...):
    for i in range(validator_count):
        ...
        yield public_key, private_key
```

ralexstokes replied:
ok this is nice

pipermerriam commented on eth2/beacon/tools/builder/initializer.py:
```diff
@@ -54,6 +106,64 @@ def create_keypair_and_mock_withdraw_credentials(
+    for index, data in enumerate(deposit_data):
+        data_to_index = deposit_data[: index + 1]
+        tree, _ = make_deposit_tree_and_root(data_to_index)
+        deposits += (create_deposit(data, tree, index, len(data_to_index)),)
```
This should also be trivial to replace with `eth_utils.to_tuple` and just `yield create_deposit(...)` on each loop.

ralexstokes replied:
also nice

pipermerriam commented on eth2/configs.py:
```diff
@@ -60,6 +62,31 @@
+    if isinstance(item, bytes):
+        return encode_hex(item)
+    else:
+        return int(item)
```
this jumped out to me as confusing.  We have an `Any` type anytime it isn't `bytes` we just blindly cast it to an integer?  Maybe worth a comment?

ralexstokes replied:
the broader context here is that i'm leaning towards reworking how this config is defined so these functions are temporary just to get the "genesis data" working for now in this PR...

i agree this may not be clear but if you look at the fields of the config they are all integers except for one `bytes`field. so on `serialization` we can be lossy.

i think defining the config as an `ssz` object is overkill but we do have a very nice way to go back/forth to `json` in `py-ssz` which i'm taking advantage of... if you can point me to something that can automatically derive this serialization code that is lighter than pulling in `py-ssz.tools` that would be great :)

pipermerriam commented on eth2/configs.py:
```diff
@@ -60,6 +62,31 @@
+    return tuple(_serialize(item) for item in config)
+
+
+def _deserialize(item: Union[int, str], typ: Any) -> Any:
```
I think you want to write this function using a `TypeVar`

```python
T = TypeVar('T')
def _deserialize(items: ..., typ: Type[T]) -> T:
    ...
```

Should let you get rid of the `Any` type.

ralexstokes replied:
was just going for something quick as i'm thinking to rework how the config is defined (and ideally get "automatic" {de,}serialization out of it)...

it seems like this `TypeVar` would fail the type checker when `typ` is `bytes` (as the return type is `str` in that case...)

pipermerriam replied:
`decode_hex` returns a `bytes` value so I believe that this should still behave correctly using `TypeVar` approach.

ralexstokes replied:
ah, you are right; i had the direction of the transformation backwards!

ralexstokes replied:
this approach caused more issues w/ `mypy` so i'm electing to just keep the `Any` type for now with the understanding that an already typed solution will land in a PR landing soon... i can prioritize this PR if the `Any` is really bothersome

pipermerriam commented on eth2/configs.py:
```diff
@@ -60,6 +62,31 @@
+
+def deserialize(config: Tuple[Union[int, str], ...]) -> Eth2Config:
+    values: Tuple[Any, ...] = ()
+    for index, (_, typ) in enumerate(Eth2Config._field_types.items()):
```
Looks like `_field_types` should be made into or exposed as a public accessor to avoid accessing the private value.

ralexstokes replied:
these are in the stdlib docs... referring to the above i think `NamedTuple` is not exactly what we want here and the serialization change for the `Eth2Config` in this PR are temporary

pipermerriam commented on trinity/components/eth2/beacon/component.py:
```diff
@@ -104,9 +67,7 @@ def is_enabled(self) -> bool:
-            privkey = Secp256k1PrivateKey.new(
-                decode_hex(boot_info.args.beacon_nodekey)
-            )
+            privkey = Secp256k1PrivateKey.new(decode_hex(boot_info.args.beacon_nodekey))
```
This module looks like the changes are mostly `black` code reformatting?  I'd suggest separating those changes from any actual changes since it isn't trivial to separate the two.  Should also reduce the total diff size of this PR to a more manageable size.

ralexstokes replied:
agreed, i didn't expect anyone to review this particular PR --- thanks for looking it over :) i'll pull out the formatting changes into a separate PR and rebase on top

pipermerriam commented on trinity/components/eth2/interop/component.py:
```diff
@@ -159,23 +126,23 @@ def munge_all_args(cls, args: Namespace, trinity_config: TrinityConfig) -> None:
-            state = state.set(
-                "genesis_time", start_time
-            )
+            state = state.set("genesis_time", start_time)
```
Same here, lots of pure formatting changes that make it hard to see what has actually changed.

ralexstokes commented on trinity/components/eth2/beacon/component.py:
```diff
@@ -2,15 +2,15 @@
 import asyncio
```
changes in this file are the beginning steps to remove the in-process validator in lieu of the validator client (which this PR is setting the ground work for)

ralexstokes commented on trinity/components/eth2/discv5/component.py:
```diff
@@ -39,8 +40,9 @@
 
-
-logger = logging.getLogger(__name__)
+logger = logging.getLogger(
```
somewhat unrelated to this PR, but the way we current have logging set up, it would just log `component` as the name of the thing dumping `discv5` logs which was very apparent when we only have the `BeaconNode` and `discv5` components.

this set of changes patches up the logger name to be more descriptive

ralexstokes commented on trinity/components/eth2/interop/component.py:
```diff
@@ -139,11 +138,11 @@ def munge_all_args(cls, args: Namespace, trinity_config: TrinityConfig) -> None:
         # Save the genesis state to the data dir!
         yaml = YAML(typ="unsafe")
-        with open(trinity_config.trinity_root_dir / GENESIS_FILE, "w") as f:
+        with open(trinity_config.trinity_root_dir / "genesis", "w") as f:
```
in line w/ the message of the commit that introduced these changes, this is temporary to clean up these constants we don't need for directories/files that won't exist anymore

ralexstokes commented on trinity/components/eth2/network_generator/component.py:
```diff
@@ -0,0 +1,113 @@
+        )
+        output_file_path.parent.mkdir(parents=True, exist_ok=True)
+        with open(output_file_path, "w") as output_file:
+            output_file.write(json.dumps(output))
```
the point of this component is to generate a **single** config file that has everything a beacon node or validator client would need to start from genesis.

ralexstokes commented on trinity/components/registry.py:
```diff
@@ -111,4 +111,4 @@ def get_components_for_eth1_client() -> Tuple[Type[BaseComponentAPI], ...]:
 
 def get_components_for_beacon_client() -> Tuple[Type[BaseComponentAPI], ...]:
-    return get_all_components(*BEACON_NODE_COMPONENTS)
+    return BEACON_NODE_COMPONENTS
```
the peer discovery component was not playing nicely and we have the discv5 component that should replace it either way -- drop down to specifically specify eth2 components for now

ralexstokes commented on trinity/config.py:
```diff
@@ -19,7 +19,6 @@
     Any,
```
changes in this file are also somewhat WIP-y but moving towards launching the beacon node from the aforementioned config file this PR generates

ralexstokes commented:
rebased away the formatting changes, added some clarifying comments that will be helpful for anyone reviewing... i'll incorporate the comments above soon and then this will be ready for a final review before merge

ralexstokes commented:
CI failure is unrelated so I'm going to move ahead w/ merge.

@pipermerriam if you want to make another pass, please do! i left some comments to clarify where i'm heading... but i would like to merge this in so other PRs on top don't become a huge chain dependent on this one

pipermerriam commented on eth2/beacon/tools/builder/initializer.py:
```diff
@@ -1,29 +1,77 @@
+    for _, validator in enumerate(genesis_state.validators):
+        public_key = validator.pubkey
+        private_key = key_pairs[public_key]
+        key_map += (
```
This is another place where the `to_tuple` + `yield` pattern would remove mutability.

pipermerriam commented on eth2/beacon/tools/builder/initializer.py:
```diff
@@ -1,29 +1,77 @@
+    for key_pair in encoded_key_map:
+        public_key = public_key_codec(key_pair["public_key"])
+        private_key = private_key_codec(key_pair["private_key"])
+        key_map[public_key] = private_key
```
This could also use the `to_dict` to remove mutability.

pipermerriam commented on eth2/beacon/tools/builder/initializer.py:
```diff
@@ -54,6 +106,62 @@ def create_keypair_and_mock_withdraw_credentials(
+def create_deposit_proof(
+    tree: MerkleTree, index: int, leaf_count: int
+) -> Tuple[Hash32, ...]:
+    length_mix_in = Hash32(leaf_count.to_bytes(32, byteorder="little"))
```
Given that this is just a `tools` module it doesn't really matter, but a while back @hwwhww figured out that there is a measurable performance difference between specifying the `byteorder` argument via keyword or positionally (with the positional approach being non-trivially faster).

ralexstokes replied:
hmmm good to know

pipermerriam commented on trinity/components/eth2/discv5/component.py:
```diff
@@ -39,8 +40,9 @@
-
-logger = logging.getLogger(__name__)
+logger = logging.getLogger(
+    re.sub(re.escape("component") + "$", "DiscV5Component", __name__)
```
This seems weird when you could just specify the path fully.  This only seems to make sense if we assume `__name__` is going to change often (such as if this file moved around a bunch).  I think just giving this a static name might be a better choice as this line isn't the easiest to grok.

ralexstokes replied:
fair enough... i do agree that this component shouldn't move much so the thrash on a static name should be minimal
